### DISCUSSION TITLE
Tasmotized devices (Sonoff) connect to AP only on 802.11b/g, not N (lite, as there is only 1 antenna)......

### DISCUSSION THREAD

**Discussion by laurentppol** [arendst/Tasmota#15584] [Q&A] [answered]
Hi there, 
here are my observations on Tasmotized (vs "other") devices  connecting to AP on TP-LINK WR1043ND (3R3T), running OpenWRT 18.06.
- Tasmotized devices connect   only on B/G: never seen more than 54Mbps, no "MCS", no 40MHz channel. Even the distance AP - Tasmota is few meters.
- "Original" (ITEAD RF Bridge ), my old phones running Android connect using N - MCS, > 54Mbps.

> [Master "adelin3"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio0.network1) (wlan1)
F4:EC:38:F2:97:E2
?
-61 / -91 dBm
39.0 Mbit/s, 20MHz, MCS 4
90.0 Mbit/s, 40MHz, MCS 4, Short GI
[Master "adelin3"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio0.network1) (wlan1)
F8:D1:11:54:47:9E
?
-67 / -91 dBm
81.0 Mbit/s, 40MHz, MCS 4
81.0 Mbit/s, 40MHz, MCS 4

The two above are TP-LINK MR3220 routers, also OpenWRT (14.07, AFAIR).
Same device (Sonoff RF Bridge R2v2.2) with Tasmota connect using B/G, using ITEAD firmware using N.
Also my own program on NodeMCU (Sming framework based, not Arduino).

> [Master "IoT"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio1.network1) (wlan0)
2C:F4:32:A9:57:B4
sonoff_4CH_pro_1 (192.168.10.234)
-65 / 0 dBm
6.0 Mbit/s, 20MHz
54.0 Mbit/s, 20MHz
[Master "IoT"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio1.network1) (wlan0)
2C:F4:32:A9:58:1B
sonoff_4CH_pro_2 (192.168.10.232)
-71 / 0 dBm
1.0 Mbit/s, 20MHz
48.0 Mbit/s, 20MHz
[Master "IoT"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio1.network1) (wlan0)
40:40:A7:62:2C:9B
192.168.10.205
-51 / 0 dBm
1.0 Mbit/s, 20MHz
72.2 Mbit/s, 20MHz, MCS 7, Short GI

And here 2 Sonoffs 4CH Pro Tasmotized, one Android, same AP card, same network.

> [Master "adelin3"](http://192.168.1.30/cgi-bin/luci/admin/network/wireless/radio0.network1) (wlan1)
EC:FA:BC:12:D2:73
NodeMCU_temometr_bimbrownika (192.168.1.201)
-57 / -91 dBm
48.0 Mbit/s, 20MHz
52.0 Mbit/s, 20MHz, MCS 5

My NodeMCU with program based on Sming framework.

So in my opinion there is something wrong with Arduino framework.
As ONLY Tasmotized (so Arduino based) devices have this problem.
Not a big problem, as none expects Tasmota to transfer big amounts of data, but existing.

**Comment by Jason2866** [comment]
Probably related https://github.com/arendst/Tasmota/discussions/13623
Lots of infos too.

**Reply by laurentppol** [reply]
What do You think: is it Arduino issue as I suggest?
My code, based on quite old SDK / framework:
-   Sming 2017-02, 
- SDK: noNos, "V2.0.0_16_08_10"

I have NEVER had problems with WiFi. But this framework has only parts of code from Arduino. Especially sensors etc. NOT the core. Just ask them on gitter.im. 
https://gitter.im/SmingHub/Sming# 

And my knowledge of C++ is very poor, so I just copied WiFi / Inet related code from examples.

**Comment by Jason2866** [comment, marked as answer]
Yes, if you read through the long discussion #13623 there is a part about the underlying NonOs SDK (closed source) which has problems with some routers. In short can not be solved.
The used NonOs SDK version for Tasmota is the only one which has no wifi reconnect issues (in our tests).
No NonOs SDK is bug free :-(  Since it is EoL this will never fixed.
If you choose the "right" AP it will connect in mode "N"

**Reply by laurentppol** [reply]
What version of SDK You use? I told about mine. Also "closed" but working. (same router!).
Could You try to make some old version of Tasmota based on same SDK I use?
To compare, if the problem is in SDK OR in Arduino.

And more: I mention about SAME hardware - RF Bridge, flashed with Tasmota, connects B/G, fIashed  ITEAD firmware connects N!
Same router!
Don't they (at ITEAD) use same SDK (maybe not exactly same version, but SDK from Espressif)??

**Reply by Jason2866** [reply]
ITEAD probably do not use Arduino nor NonOsSDK. You have to ask them what the use. Idk.
If you want to use a different NonOS SDK you have to compile yourself. 
Tasmota uses `NONOSDK22x_190703 = 2.2.2-dev(38a443e)` as default.
See https://github.com/arendst/Tasmota/blob/development/platformio.ini#L98

**Reply by laurentppol** [reply]
1). Do You know another SDK for Espressif chips? 
For Arduino I'll agree.
2). Tasmota is built on top of Arduino and Espressif SDK.
3). NOW Tasmota uses 22x_190703.
4). For me it would  be hard to find  an **old version** of Tasmota WHICH uses SDK I use.
I don't know git enough to know if it is possible to run something like grep across versions of a file.

I tried to compile, it is not hard IF You know what You want to compile (version / variant) and options :)
But a bit harder than kernel's make menuconfig. :)

Do You know some software for ESP8266 that doesn't use Arduino but uses SDK? I have time to test.

**Comment by Jason2866** [comment]
1) https://github.com/orgs/espressif/repositories?q=8266&type=all&language=&sort=
2) correct
3) as i told
4) Tasmota never used V2.0.0_16_08_10 NonOS SDK version
